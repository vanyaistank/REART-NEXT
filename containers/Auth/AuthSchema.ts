import gql from 'graphql-tag';

export const SignMutation = gql`
#	mutation($email: String!, $username: String, $password: String!) {
#		Sign(input: { email: $email, username: $username, password: $password }) {
#			success
#			message
#			data {
#				sign {
#					token
#				}
#				user {
#					username
#					ID
#					email
#				}
#			}
#		}
#	}
    mutation($email: String!, $username: String, $password: String!) {
        sign(signInput: { email: $email, username: $username, password: $password }){
            data {
                sign
                user {
                    id
                    email
                    username
                }
            }
            message
            success
        }
    }
`;

export type SignUpVariables = {
	email: string;
	username?: string;
	password: string;
};
