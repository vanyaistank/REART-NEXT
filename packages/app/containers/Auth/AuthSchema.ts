import gql from 'graphql-tag';

export const SignMutation = gql`
#	mutation($email: String!, $username: String, $password: String!) {
#		Sign(input: { email: $email, username: $username, password: $password }) {
#			success
#			message
#			data {
#				sign {
#					token
#				}
#				user {
#					username
#					ID
#					email
#				}
#			}
#		}
#	}
    mutation(
        $email: String!, 
        $username: String, 
        $password: String!, 
        $firstName: String, 
        $lastName: String
    ) {
        sign(signInput: { 
            email: $email, 
            username: $username, 
            password: $password, 
            firstName: $firstName, 
            lastName: $lastName 
        }){
            data {
                sign
                user {
                    id
                    email
                    username
	                firstName
	                lastName
                }
            }
            message
            success
        }
    }
`;

export type SignUpVariables = {
	email: string;
	username?: string;
	password: string;
	firstName?: string;
	lastName?: string;
};
