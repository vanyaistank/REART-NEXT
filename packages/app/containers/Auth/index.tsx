import React, { PureComponent } from 'react';
import { createPortal } from 'react-dom';
import { Formik, FormikProps } from 'formik';
import _ from 'lodash';

import { Mutation } from 'react-apollo';
import { SignMutation } from './AuthSchema';

import { connect } from 'react-redux';
import { getUserAndLayout, getUserAndLayoutType } from '@Redux/reselect';
import { mapDispatchToProps } from '@Redux/helpers';
import { ToggleModal } from '@Redux/reducers/Layout/LayoutActions';

import { UserData } from '@Redux/reducers/User/UserReducer';
import { Authorize } from '@Redux/reducers/User/UserActions';

import SignForm from './components/SignForm';
import {
	validateSignForm,
	possibleApiErrorMessages,
} from '@Helpers/validation/SignValidation';

export type SignFormValues = {
	email: string;
	username?: string;
	password: string;
	firstName?: string;
	lastName?: string;
};

export type oneOfFields = 'username' | 'email' | 'password' | 'firstName' | 'lastName';

interface Props extends getUserAndLayoutType {
	toggleModal: () => void;
	authorize: (data: UserData) => void; // log in user
}

interface State extends SignFormValues {
	isLogin: boolean; // whether or not show login page
	apiErrors: Partial<SignFormValues>; // wrong password and errors like that
	modalRoot: HTMLElement; // node for modal
}

interface SignUpResponseInterface {
	success: boolean;
	message?: string; // validation errors, etc
	data?: {
		sign: string;
		user: {
			username: string;
			ID: number;
			email: string;
		};
	};
}

export class AuthContainer extends PureComponent<Props, State> {
	// I will pass this to resetErrors handler when onChange event occurs to reset errors,
	// cause formik doesn't handle these
	apiErrorsInitialState: Partial<SignFormValues> = {
		username: null,
		email: null,
		password: null,
		firstName: null,
		lastName: null,
	};

	state = {
		isLogin: true,
		username: '',
		email: '',
		password: '',
		apiErrors: this.apiErrorsInitialState,
		modalRoot: null,
	};

	componentDidMount(): void {
		const modalRoot = document.getElementById('modal-root') as HTMLElement;
		this.setState({ modalRoot });
	}

	// switch between sign in and sign up screens
	switchScreen = (): void =>
		this.setState(state => ({ isLogin: !state.isLogin }));

	resetErrors = (): void =>
		this.setState({ apiErrors: this.apiErrorsInitialState });

	handleSignUpResponse = (response: SignUpResponseInterface): void => {
		const { authorize, toggleModal } = this.props;

		// since optional chaining operator isn't implemented yet in ts
		if (response) {
			if (response.success === true) {
				const { data = {} } = response;

				const hasToken = _.has(data, 'sign');
				const hasUser = _.has(data, 'user');

				if (hasToken && hasUser) {
					const payload = {
						...response.data.user,
						token: response.data.sign,
					};

					authorize(payload);
					return toggleModal();
				}
			} else if (response.success === false) {
				const { message = null } = response;

				if (message) {
					type Error = {
						error: string;
						field: oneOfFields;
					};

					// get field containing error message, e.g. "email"
					const { field = null }: Error = ({} = _.find(
						possibleApiErrorMessages,
						item => (item.error = message)
					));

					if (field) {
						this.setState(state => ({
							apiErrors: {
								...state.apiErrors,
								[field]: message,
							},
						}));
					}
				}
			}
		}
	};

	render() {
		const {
			toggleModal,
			layout: { showModal },
		} = this.props;

		const { isLogin, apiErrors, modalRoot } = this.state;

		const initialValues: SignFormValues = {
			...(isLogin === false && {
				username: '',
				firstName: '',
				lastName: '',
			}),
			email: '',
			password: '',
		};

		if (modalRoot) {
			return createPortal(
				<Mutation
					mutation={SignMutation}
					onCompleted={({ sign }) => this.handleSignUpResponse(sign)}
				>
					{(SignUpRequest, { error, loading }) => {
						// TODO: Design error component
						if (error) return <h1>Error!</h1>;

						return (
							<Formik
								initialValues={initialValues}
								onSubmit={async (variables: SignFormValues, { resetForm }) => {
									await SignUpRequest({ variables });
									return resetForm();
								}}
								validate={(values: SignFormValues) =>
									validateSignForm(values, isLogin)
								}
								render={({
									handleChange,
									...rest
								}: FormikProps<SignFormValues>) => (
									<SignForm
										apiErrors={apiErrors}
										handleChange={e => {
											this.resetErrors();
											return handleChange(e);
										}}
										toggleModal={toggleModal}
										switchScreen={this.switchScreen}
										showModal={showModal}
										loading={loading}
										{...this.state}
										{...rest}
									/>
								)}
							/>
						);
					}}
				</Mutation>,
				modalRoot
			);
		}

		return null;
	}
}

const actionCreators = {
	toggleModal: ToggleModal,
	authorize: Authorize,
};

export default connect(
	getUserAndLayout,
	mapDispatchToProps(actionCreators)
)(AuthContainer);
