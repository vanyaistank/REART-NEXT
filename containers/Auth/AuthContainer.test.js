import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import sinon from 'sinon';

import { AuthContainer } from '../Auth';

it('handles sign response successfully, then triggers toggleModal and authorize actions', () => {
	const fakeResponse = {
		success: true,
		message: null,
		data: {
			sign: {
				token: 'token',
			},
			user: {
				username: 'username!',
			},
		},
	};

	const onAuthorize = sinon.spy();
	const onToggleModal = sinon.spy();

	const wrapper = shallow(
		<AuthContainer
			toggleModal={onToggleModal}
			authorize={onAuthorize}
			user={{}}
			layout={{}}
		/>
	);

	wrapper.instance().handleSignUpResponse(fakeResponse);
	expect(onAuthorize.calledOnce).toEqual(true);
	expect(onToggleModal.calledOnce).toEqual(true);
});

it('handles error from sign response', () => {
	const fakeResponse = {
		success: false,
		message:
			'The email address and username you have entered is already registered.',
	};

	const onAuthorize = sinon.spy();
	const onToggleModal = sinon.spy();

	const wrapper = shallow(
		<AuthContainer
			toggleModal={onToggleModal}
			authorize={onAuthorize}
			user={{}}
			layout={{}}
		/>
	);

	wrapper.instance().handleSignUpResponse(fakeResponse);

	const errors = wrapper.state().apiErrors;
	expect(errors).toMatchSnapshot();
});
