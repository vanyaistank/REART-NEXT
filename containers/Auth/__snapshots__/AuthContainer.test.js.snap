// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles error from sign response 1`] = `
Object {
  "email": "The email address and username you have entered is already registered.",
  "password": null,
  "username": null,
}
`;
